Autonav

johann1764 opened the pull request:
**Bug Fix** / **New Feature**

This PR addresses _some_ of the bugs/features described in #2774.

## Summary
 - [x] If _Jump or Land to follow_ is off and If _Brake at approach_ is on, the player following a ship that is lost due to landing or jumping should stop at the  position where the target was lost.
    - [x] In order to do that, the player should ignore lanes (land).
    - [x] In order to do that, the player should ignore lanes (jump).
 - [x] If can't follow a pilot in a jump, stop autonav and issue relevant message:
    - [x] if unknown jump or exit_only;
    - [x] if out of fuel.
 - [x] Extend _"Jump to follow"_ into _"Jump or land to follow"_.
    - [x] make substitutions and update UI.
    - [x] land where the pilot landed.
 - [x]  Solve **the afterburner bug** : you get into a system with your autonav on that should redirect you to your next jump, but if you hit the afterburner too early, the ship just continues with the orientation it had while entering. This was caused by not taking afterburner accel bonus into account while testing whether current speed was unreachable by normal means.
 - [x] When we follow a pilot with jump_to_follow on an the pilot is lost, fast forward mode is stopped, as there is usually not enough time to ramp down before jumping or landing if the pilot was followed closely.
 - [x] When we follow a pilot with jump_to_follow off but brake_at_pos on, when the pilot is lost, we stop fast forward mode (ramping down does not work convincingly) and go to the last position the pilot was seen. If he has jumped (resp. landed) the jump (resp the spob) is selected.
 - [x] `autonav_abort()` with _Brake at approach_ on should stop the ship before stopping autonav.

## Testing Done
Each feature, one at a time. A validated feature has been tested.

### Don't forget
 - [x] Comment out debug messages when done.

bobbens commented on dat/events/settings.lua:
```diff
@@ -194,7 +194,7 @@ function settings ()
          var.push( "autonav_compr_max", compr_max )
          var.push( "autonav_match_fleet", match_fleet )
-         var.push( "autonav_follow_jump", follow_jump )
+         var.push( "autonav_follow_land_jump", follow_land_jump )
```
I would probably not rename the variable name, just keep it at `"autonav_follow_jump"` or we'll have to add logic to update the variable names and stuff. The local variable can be called `follow_land_jump` though.

johann1764 replied:
Do you mean update it in savegames ?

bobbens replied:
Yes. It's just a bit more trouble that we can easily avoid :)

johann1764 replied:
Sure. Next commit !

bobbens commented:
It looks good. I would remove some of the commented debug prints and such, but I think it's ready to merge. I assume this is not a draft anymore?

johann1764 commented:
I could not find a way to do the last task of the list.

I will leave a comment in the code about it, and remove the remaining debug messages.

I will also update the CHANGES.

I will mark this PR as ready when done.

johann1764 commented:
I didn't try doing a new function, as I though I might be redundant with an existing one.
But since no existing func works for that, I will try it !

bobbens commented:
> It worked!
> 
> Now I am having a doubt whether these `_("...")` I added are OK, since the function that receives it also applies `_()` on top of it. Should I replace my `_("...")` them with just `"..."` ?
> 
> If not, and if luacheck test succeeds, It should be ok.

The way gettext works is you want to apply it to the source strings, or they won't be captured in the translation software. So something like:
```lua
local foo = _("bar") -- Translated here
print( foo )  -- Translated version is used
```
Is what you want. In the case you have to use a variable for some reason or other to translate later, you should do something like:
```lua
local foo = N_("bar") -- This is actually a noop, but gets translated
print( _(foo) ) -- Now it will properly get converted
```

What would be wrong would be something like:
```lua
local foo = _("bar") -- Translated here
print( _(foo) ) -- Translated version is translated again, could become something weird...
```
The string is getting translated twice, which means that it could potentially get converted into something weird.

johann1764 commented:
Ok, just this last correction and we are good.

johann1764 commented:
Ok, fixed my mistake. 
Some instances of that mistake were not my doing, I also fixed them. (system name and pilot name could be translated twice)

Now _luacheck_ will tell whether we are ok.

bobbens commented on dat/autonav.lua:
```diff
@@ -633,8 +659,9 @@ end
+
    if player.tryLand(false)=="impossible" then
-      return autonav_abort()
+      return autonav_abort("cannot land")
```
```suggestion
      return autonav_abort(_("cannot land"))
```
Should be translated, as this is used in a format string which is translated before formatted.

bobbens commented on dat/autonav.lua:
```diff
@@ -656,24 +683,50 @@ function autonav_plt_follow ()
+               local why=nil
+               if fuel < consumption then
+                  if fuel == 0 then
+                     why="no fuel"
```
```suggestion
                     why=_("no fuel")
```

bobbens commented on dat/autonav.lua:
```diff
@@ -656,24 +683,50 @@ function autonav_plt_follow ()
+                  if fuel == 0 then
+                     why="no fuel"
+                  else
+                     why="not enough fuel"
```
```suggestion
                     why=_("not enough fuel")
```

bobbens commented on dat/autonav.lua:
```diff
@@ -732,7 +785,7 @@ function autonav_plt_board_approach ()
       autonav_end()
    elseif brd~="retry" then
-      autonav_abort()
+      autonav_abort("cannot board")
```
```suggestion
      autonav_abort(_("cannot board"))
```

bobbens commented on dat/autonav.lua:
```diff
@@ -792,8 +845,7 @@ function autonav_enter ()
       if fuel < consumption then
-         player.msg("#r".._("Autonav: not enough fuel to continue").."#0")
-         autonav_end()
+         autonav_abort("not enough fuel to continue")
```
```suggestion
         autonav_abort(_("not enough fuel to continue"))
```

bobbens reviewed (changes requested):
OK, so I mentioned double translated, this doesn't happen with format strings. So something like:
```lua
local foo = _("bar")
print( fmt.f( _("hoge {foo}"), {foo=foo} )
```
Is valid. It translates the first string and saves that as a variable, then it translates `"hoge {foo}"` where the `{foo}` part should remain unchanged, and then the translated `{foo}` gets replaced by the translated local variable string.

johann1764 commented:
I think there is a couple occurences left in "get_..._name", no time now, do it later.

johann1764 commented:
Looks like it's ok.

bobbens commented:
Looks good, thanks!
